contracts/axelarnet-gateway/src/contract/execute.rs

```diff
@@ -0,0 +1,147 @@
+        .add_event(AxelarnetGatewayEvent::ContractCalled { msg, payload }.into())
+        .add_events(events))
+}
+
+// Because the messages came from the router, we can assume they are already verified
+pub(crate) fn receive_messages(
+    store: &mut dyn Storage,
+    msgs: Vec<Message>,
+) -> Result<Response, Error> {
+    for msg in msgs.iter() {
+        state::save_outgoing_msg(store, msg.cc_id.clone(), msg.clone())
+            .change_context(Error::SaveOutgoingMessage)?;
+    }
+
+    Ok(Response::new().add_events(
+        msgs.into_iter()
+            .map(|msg| AxelarnetGatewayEvent::Routing { msg }.into()),
+    ))
+}
+
+pub(crate) fn send_messages(
+    store: &mut dyn Storage,
+    router: &Router,
+    msgs: Vec<Message>,
+) -> Result<Response, Error> {
+    for msg in msgs.iter() {
+        let stored_msg = state::may_load_incoming_msg(store, &msg.cc_id)
+            .change_context(Error::InvalidStoreAccess)?;
+
+        match stored_msg {
+            Some(message) if msg != &message => {
+                Err(report!(Error::MessageMismatch(msg.cc_id.clone())))
+            }
+            Some(_) => Ok(()),
+            None => Err(report!(Error::MessageNotFound(msg.cc_id.clone()))),
+        }?
+    }
+
+    let (wasm_msg, events) = route(router, msgs)?;
+
+    Ok(Response::new().add_message(wasm_msg).add_events(events))
+}
+
+fn route(
+    router: &Router,
+    msgs: Vec<Message>,
+) -> Result<(WasmMsg, impl IntoIterator<Item = Event>), Error> {
+    Ok((
+        router.route(msgs.clone()).ok_or(Error::RoutingFailed)?,
+        msgs.into_iter()
+            .map(|msg| AxelarnetGatewayEvent::Routing { msg }.into()),
+    ))
+}
+
+pub(crate) fn execute(
+    store: &mut dyn Storage,
+    api: &dyn Api,
+    querier: QuerierWrapper,
+    chain_name: ChainName,
+    cc_id: CrossChainId,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    let msg = state::update_msg_status(store, cc_id.clone())
+        .change_context(Error::MessageStatusUpdateFailed(cc_id))?;
+
+    let payload_hash: [u8; 32] = Keccak256::digest(payload.as_slice()).into();
+    if payload_hash != msg.payload_hash {
+        return Err(report!(Error::PayloadHashMismatch));
+    }
+
+    if chain_name != msg.destination_chain {
+        return Err(report!(Error::InvalidDestinationChain(
+            msg.destination_chain
+        )));
+    }
+
+    let destination_contract = api
+        .addr_validate(&msg.destination_address)
+        .change_context(Error::InvalidAddress(msg.destination_address.to_string()))?;
+
+    let executable: AxelarExecutableClient =
+        client::Client::new(querier, destination_contract).into();
+
+    // Call the destination contract
+    // Apps are required to expose AxelarExecutableMsg::Execute interface
+    Ok(Response::new()
+        .add_message(executable.execute(msg.cc_id.clone(), msg.source_address.clone(), payload))
+        .add_event(AxelarnetGatewayEvent::MessageExecuted { msg }.into()))
+}
```

Comment:
shouldn't this file have some tests?

Reply:
yep, planning to add test coverage today